adapter,compute: COPY TO response protocol

moulimukherjee opened the pull request:
<!--
Describe the contents of the PR briefly but completely.

If you write detailed commit messages, it is acceptable to copy/paste them
here, or write "see commit messages for details." If there is only one commit
in the PR, GitHub will have already added its commit message above.
-->

Adding plumbing to send a tag response back to a COPY TO command from the s3 oneshot sink which will look like `COPY <num_of_rows_copied>`. 

Currently this does not do any real updates to s3 yet, but rather just gives us the raw row count.

The implementation is a mix of existing subscribe and peek. We don't need to provide ongoing updates like subscribe, but rather a final count of the rows copied (hence somewhat similar to peek).

### Motivation
This PR has the next set of changes for https://github.com/MaterializeInc/materialize/issues/7256

<!--
Which of the following best describes the motivation behind this PR?

  * This PR fixes a recognized bug.

    [Ensure issue is linked somewhere.]

  * This PR adds a known-desirable feature.

    [Ensure issue is linked somewhere.]

  * This PR fixes a previously unreported bug.

    [Describe the bug in detail, as if you were filing a bug report.]

  * This PR adds a feature that has not yet been specified.

    [Write a brief specification for the feature, including justification
     for its inclusion in Materialize, as if you were writing the original
     feature specification.]

   * This PR refactors existing code.

    [Describe what was wrong with the existing code, if it is not obvious.]
-->

### Tips for reviewer

Added comments inline in the PR

<!--
Leave some tips for your reviewer, like:

    * The diff is much smaller if viewed with whitespace hidden.
    * [Some function/module/file] deserves extra attention.
    * [Some function/module/file] is pure code movement and only needs a skim.

Delete this section if no tips.
-->

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [x] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  <!-- Reference the design in the description. -->
- [x] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [x] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  <!-- Ask in #team-cloud on Slack if you need help preparing the cloud PR. -->
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - <!-- Add release notes here or explicitly state that there are no user-facing behavior changes. -->

moulimukherjee commented on src/environmentd/src/http/sql.rs:
```diff
@@ -1279,6 +1278,29 @@ async fn execute_stmt<S: ResultSender>(
         },
+        ExecuteResponse::Copying { rx, .. } => {
+            let num_rows = match sender.await_rows(client.canceled(), rx).await? {
+                Ok(CopyToResponseUnary::Rows(rows)) => {
```
cc @teskje I know we talked about removing the `CopyToResponseUnary` and reusing `CopyToResponse`, but here in environmentd `CopyToResponse` is not available as it does not depend upon compute-client

moulimukherjee replied:
done

moulimukherjee commented on src/pgwire/src/protocol.rs:
```diff
@@ -2056,6 +2082,68 @@ where
+
+        self.conn.flush().await?;
+
+        let tag = format!("COPY {}", row_count);
+        self.send(BackendMessage::CopyDone).await?;
+        self.send(BackendMessage::CommandComplete { tag }).await?;
+        Ok((
+            State::Ready,
+            SendRowsEndedReason::Success {
+                rows_returned: row_count,
+            },
+        ))
```
cc @mjibson Does this response look okay to you? Also, covered in test [here](https://github.com/MaterializeInc/materialize/pull/25042/files#diff-5224c00f6dd6f3b7c98338d2b3a4973c53c0c2abfe1ed9d1487fb2d5ff1f1b14R22)

moulimukherjee replied:
This is now removed

moulimukherjee commented on test/sqllogictest/copy_to.slt:
```diff
@@ -0,0 +1,48 @@
+statement ok
+CREATE CONNECTION aws_conn TO AWS (ASSUME ROLE ARN 'assume-role', ASSUME ROLE SESSION NAME 'session-name');
+
+# Test correct tag is returned
+# TODO (mouli): currently this fails with actually: Values(["unexpected message from server"]), fix this
+simple conn=mz_system,user=mz_system
+COPY (VALUES (1, 2, 3)) TO 's3://path' WITH (AWS CONNECTION = aws_conn, FORMAT = 'csv', MAX FILE SIZE = '100MB')
+----
+COPY 2
```
I am not entirely sure how to go about debugging this. Getting "unexpected message from server" back, but I don't see that string in our codebase?

moulimukherjee commented on src/pgwire/src/protocol.rs:
```diff
@@ -1424,6 +1425,32 @@ where
+                        // We consider that to be a cancelation, rather than a query error.
+                        (Err(e), StatementEndedExecutionReason::Canceled)
+                    }
+                    Ok((state, SendRowsEndedReason::Success { rows_returned })) => (
+                        Ok(state),
+                        StatementEndedExecutionReason::Success {
+                            rows_returned: Some(rows_returned),
+                            execution_strategy: None,
+                        },
+                    ),
```
Not entirely sure about this part. Cargo culted portions from subscribe below.

moulimukherjee replied:
this has been removed

benesch commented on src/adapter/src/copy_to.rs:
```diff
@@ -0,0 +1,54 @@
+        if let Some(channel) = self.sender.take() {
+            let _ = channel.send(resp);
+        }
+    }
```
See the comment in protocol.rs before reading this one!

Basically, I think you want to take the code in protocol.rs that aggregates up the `CopyToResponse`s and do that _here_. Instead of storing a special sender for `CopyToResponse`s in `ActiveCopyTo`, I think you just want to hang on to the `execute_ctx`, and then you can call `execute_ctx.retire(ExecuteResponse::Copied(row_count))` when you've seen all the `CopyToResponse`s from the dataflow layer that you expect. Does that make sense?

moulimukherjee replied:
@benesch This portion is where adapter is receiving a single and already aggregated CopyToResponse from the controller, and then passing it along to the oneshot receiver in `ExecuteResponse::Copying { rx ...` which was earlier returned after shipping the dataflow https://github.com/MaterializeInc/materialize/pull/25042/files#diff-9a169fe3ef85a53dbf876b6b69d8d4f6f6c78901afcdf2683f72066e924b9df2R1032-R1035 (and environmentd is await-ing on it below before returning to the user).

If I eventually want to send something like `ExecuteResponse::Copied(rows)` back, what should I return after shipping the dataflow then instead of `ExecuteResponse::Copying { rx...`? Because I think we still want to send something back to the user which blocks until we are done and we receive the final row count.

moulimukherjee replied:
Oh do you mean that we not return `ExecuteResponse::Copying { rx...` from sequencer but rather wait for the dataflow responses back there and directly return `ExecuteResponse::Copied`?

moulimukherjee replied:
Have updated the PR to send a `ExecuteResponse::Copied` directly.

benesch commented on src/environmentd/src/http/sql.rs:
```diff
@@ -1279,6 +1278,29 @@ async fn execute_stmt<S: ResultSender>(
             ),
             ctx_extra,
         },
+        ExecuteResponse::Copying { rx, .. } => {
+            let num_rows = match sender.await_rows(client.canceled(), rx).await? {
+                Ok(CopyToResponseUnary::Rows(rows)) => {
+                    rows
+                }
+                Ok(CopyToResponseUnary::Error(e)) => {
+                    return Ok(
+                        SqlResult::err(client, Error::Unstructured(anyhow!(e))).into(),
+                    );
+                }
+                Ok(CopyToResponseUnary::Canceled) => {
+                    return Ok(SqlResult::err(client, AdapterError::Canceled).into());
+                }
+
+                Err(e) => {
+                    return Ok(
+                        SqlResult::err(client, Error::Unstructured(anyhow!(e))).into(),
+                    );
+                },
+            };
+            let tag = format!("COPY {}", num_rows);
+            SqlResult::ok(client, tag, Vec::default()).into()
+        }
```
Mentioned this briefly on Slack, but I think you probably _don't_ want to plumb the copy response into the pgwire protocol layer itself. Instead I think you want to aggregate the responses in the adapter layer, in the `ActiveCopyTo` processing, and just plumb through an `ExecuteResponse` to here.

moulimukherjee replied:
I pushed an update, directly sending back an `ExecuteResponse::Copied` instead of this custom thing.

benesch commented:
Another way of summarizing my comments, if helpful:

  * `COPY TO S3` response handling should look very similar to `SUBSCRIBE` response handling _from clusterd, through the controller, up to the adapter layer_.
  * Because `COPY TO S3` does not need stream results to the client but only needs to send back "COPIED <count>", the communication between the adapter layer and protocol layer should look more like the handling of, say, `INSERT` rather than `SUBSCRIBE` (or even `SELECT`).

def- commented on src/adapter/src/coord/ddl.rs:
```diff
@@ -562,6 +600,24 @@ impl Coordinator {
                 self.drop_compute_sinks(sink_ids);
             }
+            if !copy_to_sinks_to_drop.is_empty() {
+                let mut sink_ids = Vec::new();
+                for (_conn_id, dropped_name, sink_id) in copy_to_sinks_to_drop {
+                    if let Some(copy_to) = self.remove_active_copy_to(sink_id.global_id).await {
+                        self.retire_execution(
+                            StatementEndedExecutionReason::Canceled,
+                            Default::default(), // TODO (mouli): is ctx_extra required here
+                        );
+                        let _ = copy_to.sender.expect("oneshot channel exists").send(
+                            CopyToResponseUnary::Error(format!(
+                                "query could not complete because {dropped_name} was dropped"
+                            )),
+                        );
+                    }
+                    sink_ids.push(sink_id);
+                }
+                self.drop_compute_sinks(sink_ids);
```
This whole block seems uncovered?

moulimukherjee replied:
let me add more tests

moulimukherjee replied:
Added a test which should now test out this part.

def- commented on src/environmentd/src/http/sql.rs:
```diff
@@ -1279,6 +1278,29 @@ async fn execute_stmt<S: ResultSender>(
             ),
             ctx_extra,
         },
+        ExecuteResponse::Copying { rx, .. } => {
+            let num_rows = match sender.await_rows(client.canceled(), rx).await? {
+                Ok(CopyToResponseUnary::Rows(rows)) => {
+                    rows
+                }
+                Ok(CopyToResponseUnary::Error(e)) => {
+                    return Ok(
+                        SqlResult::err(client, Error::Unstructured(anyhow!(e))).into(),
+                    );
+                }
+                Ok(CopyToResponseUnary::Canceled) => {
+                    return Ok(SqlResult::err(client, AdapterError::Canceled).into());
+                }
+
+                Err(e) => {
+                    return Ok(
+                        SqlResult::err(client, Error::Unstructured(anyhow!(e))).into(),
+                    );
+                },
+            };
+            let tag = format!("COPY {}", num_rows);
+            SqlResult::ok(client, tag, Vec::default()).into()
+        }
```
Uncovered

moulimukherjee replied:
This part is now removed

def- reviewed (dismissed):
Currently the tests don't seem to be covering some interesting code, marked it inline, based on https://buildkite.com/materialize/coverage/builds/350

teskje commented on src/compute-client/src/service.rs:
```diff
@@ -163,6 +164,16 @@ pub struct PartitionedComputeState<T> {
+    /// The compute protocol requires that exactly one response is emitted for each COPY TO command. This
+    /// property ensures that a) we can eventually drop the tracking state maintained for a copy
+    /// and b) we won't re-initialize tracking for a copy we have already served.
+    pending_copy_tos: BTreeMap<GlobalId, BTreeMap<usize, CopyToResponse>>,
```
```suggestion
    copy_to_responses: BTreeMap<GlobalId, BTreeMap<usize, CopyToResponse>>,
```
For consistency with `peek_responses`.

teskje replied:
Also, a nit: Can we move this field after `peek_responses`? And also move the `CopyToResponse` handling below after the `PeekResponse` handling? That will make it easier to see that these are similar, and easier to avoid that they become inconsistent with each other.

teskje commented on src/compute-client/src/service.rs:
```diff
@@ -454,6 +468,37 @@ where
+                            (CopyToResponse::Rows(r1), CopyToResponse::Rows(r2)) => {
+                                // After the s3 sink has finished processing, each worker is
+                                // expected to send a `CopyToResponse` back, which we are aggregating
+                                // here.
```
I would drop this comment. It's pretty obvious (at least if you are familiar with the code).

If you want to keep it, can you replace "worker" with "shard"? Shards are not always workers, they might also be processes in multi-process replicas.

teskje commented on src/compute/src/sink/s3_oneshot.rs:
```diff
@@ -27,18 +32,108 @@ where
+                    response_protocol.send_final_count(row_count);
+                }
+            }
+        }
```
Fine for this PR, but I'd just like to point out that this code is prone to write data to S3 that's at times after the sink `as_of`.

Unless we assume that we are in a single-time dataflow, in which case we should probably assert that somewhere during dataflow rendering. And input draining is also not necessary in this case I think.

moulimukherjee replied:
Ack. I will add some asserts when implementing the sink properly.

teskje commented on src/compute/src/sink/s3_oneshot.rs:
```diff
@@ -27,18 +32,108 @@ where
+struct ResponseProtocol {
+    pub sink_id: GlobalId,
+    pub response_buffer: Option<Rc<RefCell<Vec<(GlobalId, CopyToResponse)>>>>,
+}
```
This should implement `Drop` to ensure a `Dropped` response is sent if (and only if) no other response was sent previously.

moulimukherjee replied:
Added! Also, since the handle to the buffer is cleared after sending one message, the `Dropped` response will not be sent if a message was sent previously.

moulimukherjee commented on src/compute-client/src/protocol/response.rs:
```diff
@@ -290,6 +316,42 @@ impl Arbitrary for PeekResponse {
+#[derive(Debug, Clone, Eq, PartialEq, Serialize, Deserialize)]
+pub enum CopyToResponse {
+    /// Returned number of rows for a successful COPY TO.
+    Rows(u64),
```
done

moulimukherjee commented on src/compute-client/src/metrics.rs:
```diff
@@ -290,7 +299,6 @@ impl InstanceMetrics {
             .observe(duration.as_secs_f64());
     }
 }
-
```
yeah, fixed

moulimukherjee commented on src/compute-client/src/protocol/response.rs:
```diff
@@ -127,6 +127,21 @@ pub enum ComputeResponse<T = mz_repr::Timestamp> {
+    /// created previously by a [`CreateDataflow` command].
+    ///
+    /// [`CreateDataflow` command]: super::command::ComputeCommand::CreateDataflow
+    CopyToResponse(GlobalId, CopyToResponse),
```
From our chat today, renaming the sink to `CopyToS3Oneshot` from `S3Oneshot`. The `CopyToResponse` buffer and all the plumbing should be re-usable if in case we add more compute sinks say like GCP or Azure's blob storage.

moulimukherjee replied:
I will actually do the rename in a separate PR to reduce the diffs in this PR.

moulimukherjee commented:
Converting to draft to fix builds

moulimukherjee commented on src/compute-client/src/controller/instance.rs:
```diff
@@ -154,6 +156,13 @@ pub(super) struct Instance<T> {
+    /// dataflows created through [`ActiveInstance::create_dataflow`].
+    ///
+    /// The entry for a copy to is removed once at least one replica has finished.
+    copy_tos: BTreeMap<GlobalId, Option<ReplicaId>>,
```
Removed!

moulimukherjee commented on src/compute-client/src/controller/instance.rs:
```diff
@@ -1455,6 +1455,9 @@ where
             ComputeResponse::PeekResponse(uuid, peek_response, otel_ctx) => {
                 self.handle_peek_response(uuid, peek_response, otel_ctx, replica_id)
             }
+            ComputeResponse::CopyToResponse(id, response) => {
+                Some(ComputeControllerResponse::CopyToResponse(id, response))
+            }
```
@teskje Does this look fine to you? Removed the code which was targeting the replica, so it's going to send back responses across all of them.

teskje replied:
Ah, you still need to make sure that you only return the first replica's response! What the replica-targeting does is that it only returns the response from a selected replica and ignore all others. We don't want that but the untargeted behavior which is returning the response from the fastest replica and ignoring all others.

So we still want the `copy_tos` field, but it should be a `BTreeSet<GlobalId>` (the values in the `BTreeMap` were the targeted replicas).

moulimukherjee replied:
Ah that makes much more sense and I misunderstood, updating.

moulimukherjee replied:
done!

moulimukherjee commented on src/compute-client/src/controller/instance.rs:
```diff
@@ -786,6 +809,18 @@ where
+                copy_to_id,
+                CopyToResponse::Error("target replica failed or was dropped".into()),
+            );
+            self.compute.deliver_response(response);
```
this has now been removed

moulimukherjee commented:
Have addressed all the comments. I think this is now in a good place for another round of reviews.

moulimukherjee commented on test/sqllogictest/copy_to.slt:
```diff
@@ -0,0 +1,46 @@
+simple conn=mz_system,user=mz_system
+COPY (VALUES (1, 2, 3)) TO 's3://path' WITH (AWS CONNECTION = aws_conn, FORMAT = 'csv', MAX FILE SIZE = '100MB')
+----
+COMPLETE 1
```
For some reason sqllogic is returning a `COMPLETE` instead of a `COPY` tag (while pgwire tests above return the `COPY` tag). Is this expected?

moulimukherjee replied:
An `INSERT` statement also returns a `COMPLETE` tag, so looks like this is expected.

moulimukherjee commented on src/adapter/src/copy_to.rs:
```diff
@@ -0,0 +1,54 @@
+    /// All `GlobalId`s that the copy to's expression depends on.
+    pub depends_on: BTreeSet<GlobalId>,
+    /// Whether we are already in the process of dropping the resources related to this copy to.
+    pub dropping: bool,
```
Ack. Keeping the subscribe similar bits as of now.

benesch commented on src/compute-client/src/controller/instance.rs:
```diff
@@ -1536,6 +1557,20 @@ where
+        response: CopyToResponse,
+    ) -> Option<ComputeControllerResponse<T>> {
+        // We might not be tracking this COPY TO because we have already returned a response.
+        // In that case, we ignore the response.
```
This is in the case of another replica having already completed the copy?

moulimukherjee replied:
Yes! Updating the comment to clarify

benesch commented on src/adapter/src/coord/sequencer/inner/peek.rs:
```diff
@@ -825,31 +834,80 @@ impl Coordinator {
+            // after receiving one response. The unfinished sinks will emit a `Dropped` message, but
+            // these will not be sent to the adapter because a message was sent previously (the controller
+            // will only send one CopyToResponse for one corresponding COPY TO)
+            // TODO (mouli): should we panic! instead?
```
I'd recommend using types to enforce this statically if possible. It looks to me like the layer of the stack that "swallows" `CopyToResponse::Dropped` (I think the controller in `compute-client`) could convert the underlying `CopyToResponse` to a different type that doesn't have a `Dropped` variant.

On a quick skim, it looks like `ComputeResponse::CopyToResponse(GlobalId, CopyToResponse)` potentially wants to be `ComputeResponse::CopyToResponse(GlobalId, Result<u64, Error>)`?

moulimukherjee replied:
ack, makes sense, updating

moulimukherjee replied:
Updated. Converting to `Result<u64, Error>` in the `compute-client`. 

The `ComputeResponse::CopyToResponse` is still `(GlobalId, CopyToResponse)` because that's what gets aggregated in compute's `PartitionedState`, but the `ComputeControllerResponse::CopyToResponse` and then the `ControllerResponse::CopyToResponse` is using `Result`.

benesch commented on src/adapter/src/coord/sequencer/inner/peek.rs:
```diff
@@ -825,31 +834,80 @@ impl Coordinator {
+        // Ship dataflow.
+        self.ship_dataflow(df_desc, cluster_id).await;
+        let internal_cmd_tx = self.internal_cmd_tx.clone();
+        // Awaiting the response in a separate task to unblock
+        // further processing.
+        mz_ore::task::spawn(|| "copy_to await response", async move {
+            let response = match rx.await {
+                Ok(res) => Some(res),
+                Err(error) => {
+                    warn!("error receiving response back from sink: {:?}", error);
+                    None
+                }
+            };
+            let _ = internal_cmd_tx.send(Message::PeekStageReady {
+                ctx,
+                otel_ctx: root_otel_ctx,
+                stage: PeekStage::CopyToFinish(PeekStageCopyToFinish { validity, response }),
+            });
+        });
```
I'm less familiar with the adapter codebase than I used to be, but this task seems potentially unnecessary to me.

Instead of having the `ActiveCopyTo` have a `sender` that sends back to this task, which then creates a new peek stage, which then finally retires the `ExecuteContext` ... can you instead just put the `ExecuteContext` inside of `ActiveCopyTo`, and have `ActiveCopyTo` call `ctx.retire` as appropriate?

moulimukherjee replied:
that works (and it simplified the code)!

moulimukherjee commented on src/compute-client/src/controller/instance.rs:
```diff
@@ -1536,6 +1557,30 @@ where
+        // from one of the replicas. In that case, we ignore the response.
+        if self.compute.copy_tos.remove(&sink_id) {
+            let result = match response {
+                CopyToResponse::RowCount(count) => Ok(count),
+                CopyToResponse::Error(error) => Err(anyhow::anyhow!(error)),
+                // We do not expect to receive a `Dropped` message here.
+                // A sink is dropped when a response is already received.
+                // And since we can receive only one response, the `Dropped`
+                // never reaches here.
+                // TODO (mouli): should we panic! instead?
+                CopyToResponse::Dropped => Err(anyhow::anyhow!("unexpected error")),
+            };
+            Some(ComputeControllerResponse::CopyToResponse(sink_id, result))
```
@benesch Doing the `CopyToResponse` to `Result` conversion here.

benesch replied:
Nice. This seems slightly cleaner to me since the weirdness around `CopyToResponse::Dropped` is handled all internally to the controller, and the adapter never sees any of the weirdness. But I defer to @teskje for the final assessment, and also for answering the question in your TODO!

teskje replied:
The previous version where `Dropped` responses were reported to the caller had the property of being consistent with subscribe handling, so a lot of subscribe code could just be copied and adapted. I don't mind simplifying things, but then we should make extra sure that adapter also handles this response protocol correctly.

For adapter to be able to handle `CopyToResponse`s correctly, we need to agree on what the controller promises about them. I'd suggest the following:

1. For each installed copy-to, the controller will produce _at most_ one response.
2. Once a copy-to was dropped, the controller will not produce responses for it anymore.

With this contract, we'd need to change only one things: In `drop_collections`, remove all dropped IDs from `copy_tos`. Doing this guarantees that we don't return more responses for dropped copy-to. It also makes erroring on a `Dropped` response in `handle_copy_to_response` correct.

teskje replied:
And in adapter, we should make sure we don't leak state associated with the copy-to when it is canceled before it completes. I think the current adapter-side handling is modeled after subscribe, and while I don't fully grok adapter's subscribe handling, I think it keeps around subscribe state after the `drop_collections` call until it receives a final response?

benesch replied:
Btw, I had added supporting cancellation as a public preview blocker as a few days ago. The ideal case is that it's easy to rig up now, but I also think it's okay if we don't get the cancellation semantics _quite_ right while we're still iterating. (Memory leaks are a bit of a gray area though; if we release to private preview with a way for users to leak memory in `envd`, that can put the stability of the whole environment at risk.)

![image](https://github.com/MaterializeInc/materialize/assets/882976/447a7fac-6adf-44a2-8b5c-85f26197a4c3)

moulimukherjee replied:
Changed it to match the updated subscribe (returning `None` instead of `Err` for the `Dropped` case). This block is within a `if self.compute.copy_tos.remove(&sink_id) {`, so the state will be removed as soon as we receive one response.

benesch commented on src/adapter/src/copy_to.rs:
```diff
@@ -0,0 +1,52 @@
+impl ActiveCopyTo {
+    /// Processes a copy to response.
+    pub(crate) fn process_response(&mut self, result: Result<u64, anyhow::Error>) {
+        // Using an option to get an owned value after take
+        // to call `retire` on it.
+        if let Some(ctx) = self.ctx.take() {
```
Could `process_response` instead take `self` (i.e., by value)? I think that would avoid the `Option` around `ctx`, and also make it harder to forget to remove `ActiveCopyTo` from the `active_copy_tos` map when you receive the first response.

moulimukherjee replied:
Oh I see. Makes sense, let me update.

moulimukherjee replied:
The way the code's written now, we process the response, and then drop the compute sinks (which again fetches the entry from `active_copy_tos` map), and update some metrics (which reads from `active_copy_tos` as well). So, if I consume the object in `process_response`, the sink does not get dropped (and when dropping the cluster in my test it panics). I could refactor it, but subscribe also uses these methods so does not seem worth the effort? Keeping as it is now.

benesch replied:
It seems to me you could push the call to `process_response` into the call to `remove_active_copy_to`, where you've got the owned reference?

```diff
diff --git a/src/adapter/src/coord/message_handler.rs b/src/adapter/src/coord/message_handler.rs
index 93719b1e8f..e8c1cbafa3 100644
--- a/src/adapter/src/coord/message_handler.rs
+++ b/src/adapter/src/coord/message_handler.rs
@@ -368,13 +368,12 @@ impl Coordinator {
             }
             ControllerResponse::CopyToResponse(sink_id, response) => {
                 if let Some(active_copy_to) = self.active_copy_tos.get_mut(&sink_id) {
-                    active_copy_to.process_response(response);
                     let csid = ComputeSinkId {
                         cluster_id: active_copy_to.cluster_id,
                         global_id: sink_id,
                     };
                     self.drop_compute_sinks([csid]);
-                    self.remove_active_copy_to(sink_id);
+                    self.remove_active_copy_to(sink_id, response);
                 }
             }
             ControllerResponse::ComputeReplicaMetrics(replica_id, new) => {
diff --git a/src/adapter/src/coord/sequencer/inner/peek.rs b/src/adapter/src/coord/sequencer/inner/peek.rs
index 0be77f4f5b..1daf528439 100644
--- a/src/adapter/src/coord/sequencer/inner/peek.rs
+++ b/src/adapter/src/coord/sequencer/inner/peek.rs
@@ -844,7 +844,7 @@ impl Coordinator {
         let active_copy_to = ActiveCopyTo {
             user: ctx.session().user().clone(),
             conn_id: ctx.session().conn_id().clone(),
-            ctx: Some(ctx),
+            ctx,
             cluster_id,
             depends_on: validity.dependency_ids.clone(),
             dropping: false,
diff --git a/src/adapter/src/coord/sql.rs b/src/adapter/src/coord/sql.rs
index c8234f8a17..4bcd1d65a3 100644
--- a/src/adapter/src/coord/sql.rs
+++ b/src/adapter/src/coord/sql.rs
@@ -270,13 +270,14 @@ impl Coordinator {
 
     /// Handle removing metadata associated with a COPY TO query.
     #[tracing::instrument(level = "debug", skip(self))]
-    pub(crate) fn remove_active_copy_to(&mut self, id: GlobalId) {
+    pub(crate) fn remove_active_copy_to(&mut self, id: GlobalId, response: Result<u64, anyhow::Error>) {
         if let Some(active_copy_to) = self.active_copy_tos.remove(&id) {
             let session_type = metrics::session_type_label_value(&active_copy_to.user);
             self.metrics
                 .active_copy_tos
                 .with_label_values(&[session_type])
                 .dec();
+            active_copy_to.process_response(response);
         }
     }
 }
diff --git a/src/adapter/src/copy_to.rs b/src/adapter/src/copy_to.rs
index 8e7ff72ad8..f5a1ed3935 100644
--- a/src/adapter/src/copy_to.rs
+++ b/src/adapter/src/copy_to.rs
@@ -27,7 +27,7 @@ pub struct ActiveCopyTo {
     pub conn_id: ConnectionId,
     /// The context about the COPY TO statement getting executed.
     /// Used to eventually call `ctx.retire` on.
-    pub ctx: Option<ExecuteContext>,
+    pub ctx: ExecuteContext,
     /// The cluster that the copy to is running on.
     pub cluster_id: ClusterId,
     /// All `GlobalId`s that the copy to's expression depends on.
@@ -38,15 +38,11 @@ pub struct ActiveCopyTo {
 
 impl ActiveCopyTo {
     /// Processes a copy to response.
-    pub(crate) fn process_response(&mut self, result: Result<u64, anyhow::Error>) {
-        // Using an option to get an owned value after take
-        // to call `retire` on it.
-        if let Some(ctx) = self.ctx.take() {
-            let response = match result {
-                Ok(n) => Ok(ExecuteResponse::Copied(n.try_into().expect("u64 to usize"))),
-                Err(error) => Err(AdapterError::Unstructured(error)),
-            };
-            let _ = ctx.retire(response);
-        }
+    pub(crate) fn process_response(self, result: Result<u64, anyhow::Error>) {
+        let response = match result {
+            Ok(n) => Ok(ExecuteResponse::Copied(n.try_into().expect("u64 to usize"))),
+            Err(error) => Err(AdapterError::Unstructured(error)),
+        };
+        let _ = self.ctx.retire(response);
     }
 }
```

I didn't actually test this, so it's possible it still panics, but it seemed like it'd let you keep the control flow more or less the same as `SUBSCRIBE`, while still taking ownership of the `ActiveCopyTo` while processing the response.

moulimukherjee replied:
@benesch It does not feel right to me to process the request in a remove wrapper call (where it was just removing an entry and decrementing a counter)? I guess I can rename the method from `remove_active_copy_to` to `process_active_copy_to` which also removes the entry.

benesch replied:
Yeah, exactly! Could even call it “process_and_remove_active_copy_to”, or “retire_active_copy_to”? I like the latter the best I think.

moulimukherjee replied:
updated

benesch reviewed (commented):
The new handling of responses looks wonderfully simple!

benesch commented on src/adapter/src/coord/ddl.rs:
```diff
@@ -756,15 +789,21 @@ impl Coordinator {
             // Note: Ideally we'd use .filter(...) on the iterator, but that would
             // require simultaneously getting an immutable and mutable borrow of self.
             let need_to_drop = self
                 .active_subscribes
                 .get(&sink.global_id)
                 .map(|meta| !meta.dropping)
-                .unwrap_or(false);
+                .unwrap_or(false)
+                || self
+                    .active_copy_tos
+                    .get(&sink.global_id)
+                    .map(|meta| !meta.dropping)
+                    .unwrap_or(false);
```
This is a little hard to read like this. Would recommend factoring out into separate variables: `let need_to_drop_subscribe = ...; let need_to_drop_copy_to = ...; if !need_to_drop_subscribe || !need_to_drop_copy_to { ... }`.

benesch replied:
(Perhaps this disappears if you go with Parker’s suggestion of an ActiveEphemeralSink enum!)

moulimukherjee replied:
updated!

nrainer-materialize commented:
Please trigger a nightly build before merging this PR. Thank you! 😀

teskje commented on src/compute-client/src/controller/instance.rs:
```diff
@@ -1536,6 +1557,30 @@ where
+                // And since we can receive only one response, the `Dropped`
+                // never reaches here.
+                // TODO (mouli): should we panic! instead?
+                CopyToResponse::Dropped => Err(anyhow::anyhow!("unexpected error")),
```
Wait, I'm not sure that is right! What happens when the controller client (adapter) drops the sink dataflow (via `drop_collections` before the `COPY` is finished? The controller would send an `AllowCompaction([])` command to the replica, which would drop the dataflow possibly before the copy operation completed, so we would receive a `Dropped` response here.

benesch replied:
So perhaps the right thing to do is to produce an error here saying “query canceled”? I think we already have an AdapterError variant for that.

moulimukherjee replied:
This is handled similarly to the new subscribe now after https://github.com/MaterializeInc/materialize/pull/25179 with a `drop_compute_sinks` call after sending back a `AdapterError::Canceled` to the user.

ParkMyCar commented on src/adapter/src/coord.rs:
```diff
@@ -1340,6 +1345,8 @@ pub struct Coordinator {
 
     /// A map from active subscribes to the subscribe description.
     active_subscribes: BTreeMap<GlobalId, ActiveSubscribe>,
+    /// A map from active copy tos to the copy to description.
+    active_copy_tos: BTreeMap<GlobalId, ActiveCopyTo>,
```
High level comment, don't let this derail you. It seems like at least from the Adapter's perspective there are a lot of similarities between active subscribes and active copy_tos, could it simply some handling if we unify them? Maybe something like:

```
// Active dataflows that are sinking (outputting) data and are short lived, e.g.
// their lifetime is tied to a session or they have a finite amount of data.
enum ActiveEphemeralSink {
  Subscribe(...),
  CopyTo(...),
}

// ...

active_ephemeral_sinks: BTreeMap<GlobalId, ActiveEphemeralSink>,
```

moulimukherjee replied:
That would be nice! I will give it a stab

mjibson replied:
One take from reading the existing code: there's enough changing here that I would recommend doing this refactor in a followup PR to reduce the size of this PR and mental complexity of us reading what's going on. Adding this refactor will make it harder for you and the reviewers to figure out what's going on, because we'd now have to consider SUBSCRIBE.

moulimukherjee replied:
I just got done with the refactor 🙈. I would say the diffs are not too bad, rather contrasts with subscribe nicely. Since they are in separate commits, I can also spin it off as a different PR if needed.

moulimukherjee replied:
The wrapper `ActiveComputeSink` is refactored and can be reviewed in a separate commit https://github.com/MaterializeInc/materialize/pull/25042/commits/d501f03cd870ff26da70e7191da2591dcad0b3fc

ParkMyCar commented on src/adapter/src/coord/ddl.rs:
```diff
@@ -572,6 +595,16 @@ impl Coordinator {
+                let mut sink_ids = Vec::new();
+                for (conn_id, sink_id) in copy_to_sinks_to_drop {
+                    if let Some(conn_meta) = self.active_conns.get_mut(&conn_id) {
+                        conn_meta.drop_sinks.retain(|id| *id != sink_id);
```
Like subscribe, should we emit a notice for a dropped copy_to?

moulimukherjee replied:
This got refactored a lot in https://github.com/MaterializeInc/materialize/pull/25179, which is not sending a notice anymore. Will revisit after that's merged.

benesch commented on src/adapter/src/coord/command_handler.rs:
```diff
@@ -904,6 +904,27 @@ impl Coordinator {
+                    .active_copy_tos
+                    .remove(&sink_id)
+                    .expect("known to exist from call above");
+                // TODO (mouli): Should drop_compute_sinks be called here as well? `drop_compute_sinks`
```
This is the question that led me on the #25042 quest!

benesch reviewed (commented):
I tried to understand the new cancellation flow here, and got pretty thoroughly turned trying to understand the existing subscribe lifecycle. The subscribe response handling is in need of some love! I put up https://github.com/MaterializeInc/materialize/pull/25179 to try to clean it up.

I realize this is a bit of a wrench, but I'd like to at least consider merging #25179 ahead of this PR. I think it will be _much_ simpler to slot the `COPY TO` response handling in alongside the `SUBSCRIBE` response handling if we merge #25179 first. I don't think we need to wait more than a day or two, but if #25179 sails through review I think it will be worthwhile—and I'm happy to help with rebasing this PR on top of #25179. If #25179 does get stuck in review, we can reevaluate.

moulimukherjee commented:
Converting to draft till I am done rebasing and refactoring.

mjibson commented on src/adapter/src/subscribe.rs:
```diff
@@ -316,12 +369,38 @@ impl ActiveSubscribe {
+}
+
+impl ActiveCopyTo {
+    pub(crate) fn process_response(&mut self, response: Result<ExecuteResponse, AdapterError>) {
```
This function either should not exist (my preference), or take the owned `self`, making the struct unusable after calling.

moulimukherjee replied:
(Resolving since we addressed this in slack)

mjibson commented on src/adapter/src/subscribe.rs:
```diff
@@ -316,12 +369,38 @@ impl ActiveSubscribe {
+    pub conn_id: ConnectionId,
+    /// The context about the COPY TO statement getting executed.
+    /// Used to eventually call `ctx.retire` on.
+    pub ctx: Option<ExecuteContext>,
```
Very scared about having this be an option, because it obfuscates the responsibility of which component owns ctx retiring.

benesch replied:
This is my fault, because I changed subscribe to use an `Option` in the morally equivalent spot: https://github.com/MaterializeInc/materialize/pull/25179/files#diff-20ea8c38b1da4ccc88fd84efcf5a99097e62bc1811e726cee5562c9c97c4bb98R41

The idea is that `None` indicates "final response already sent to the subscribe (or copy to), move along, all that's remaining is to call `drop_collections` on the thing."

moulimukherjee replied:
(Resolving since we addressed this in slack)

moulimukherjee commented on test/sqllogictest/copy_to.slt:
```diff
@@ -0,0 +1,54 @@
+COMPLETE 2
+
+simple conn=mz_system,user=mz_system
+COPY t TO 's3://path' WITH (AWS CONNECTION = aws_conn, FORMAT = 'csv', MAX FILE SIZE = '100MB')
```
There's nothing written to s3 yet 😬  (will add in follow up PRs)

moulimukherjee commented on src/adapter/src/coord/message_handler.rs:
```diff
@@ -351,6 +351,15 @@ impl Coordinator {
                 }
             }
+            ControllerResponse::CopyToResponse(sink_id, response) => {
+                if let Some(active_copy_to) = self.active_copy_tos.get_mut(&sink_id) {
```
The reason it's not, because we still need the state here, because `drop_compute_sinks` will be called later via `clear_connection` and that method drops the compute collections (it needs this entry to be available there) and then clears the state.

moulimukherjee replied:
(Resolving since we addressed this in slack)

mjibson commented:
Whatever works best for you with the subscribe refactor is fine. Separate commit, or even separate PR before or after this is fine with me.

benesch commented on src/adapter/src/subscribe.rs:
```diff
@@ -27,6 +27,59 @@ use timely::progress::Antichain;
 use tokio::sync::mpsc;
```
We should rename this file to `compute_sink.rs` and adjust the comment on L10 accordingly.

moulimukherjee replied:
It would show up in the diffs as an entirely new file then. I thought it might be harder to review. Once I get an approval, I will do the rename as the last step before merging?

moulimukherjee replied:
done

benesch commented on src/adapter/src/subscribe.rs:
```diff
@@ -316,12 +369,38 @@ impl ActiveSubscribe {
+#[derive(Debug)]
+pub struct ActiveCopyTo {
+    /// The user of the session that created the subscribe.
+    pub user: User,
```
They're needed because `ctx` can become `None` when `process_response` is called, but the drop code is still be interested in the `ActiveCopyTo`s connection and user ID.

benesch commented:
I think I see a really nice refactor to avoid the `Option` in `ActiveCopyTo` and `ActiveSubscribe`. Let me try to sketch it out.

benesch commented:
> I think I see a really nice refactor to avoid the `Option` in `ActiveCopyTo` and `ActiveSubscribe`. Let me try to sketch it out.

@mjibson @moulimukherjee here's the refactor, based on current main: https://github.com/MaterializeInc/materialize/pull/25260

I'm _really_ leery of delaying this PR any further by sliding another PR in on main that will force @moulimukherjee to refactor here. But it'd be good to try to work the idea in somehow. Maybe the best course of action, @moulimukherjee, would be for you to cherry pick #25260 into this PR? I think it will slot in _really_ nicely next to the refactor you did. It will let you adjust `ActiveCopyTo::process_response` to take `self` by value, as @mjibson requested, which should make the semantics of `ctx` much clearer!

benesch commented on src/adapter/src/subscribe.rs:
```diff
@@ -27,6 +27,59 @@ use timely::progress::Antichain;
+pub enum ActiveComputeSink {
+    Subscribe(ActiveSubscribe),
+    CopyTo(ActiveCopyTo),
+}
```
I really like how this turned out!

moulimukherjee commented on src/compute-client/src/controller/instance.rs:
```diff
@@ -1541,6 +1562,37 @@ where
+        }
+    }
+
     fn handle_subscribe_response(
```
Yeah, this entire logic is within a `self.compute.copy_tos.remove` above.

moulimukherjee replied:
updated!

teskje replied:
You still need to call `copy_tos.remove` in `drop_collections` though! Right now the assumption stated in the comment above is not true ("when the controller drops a copy to it also removes it from the list of tracked copy_tos") and we will always get an error logged when a copy-to is canceled.

moulimukherjee replied:
> You still need to call copy_tos.remove in drop_collections
Just pushed the change for that

moulimukherjee commented on src/adapter/src/subscribe.rs:
```diff
@@ -27,6 +27,59 @@ use timely::progress::Antichain;
+    pub fn process_subscribe(&mut self, batch: SubscribeBatch) {
+        match self {
+            ActiveEphemeralSink::Subscribe(subscribe) => subscribe.process_response(batch),
+            ActiveEphemeralSink::CopyTo(_) => {}
```
updated!

teskje commented on src/compute-client/src/controller/instance.rs:
```diff
@@ -154,6 +156,13 @@ pub(super) struct Instance<T> {
+    /// New entries are added for all s3 oneshot sinks (corresponding to a COPY TO) exported from
+    /// dataflows created through [`ActiveInstance::create_dataflow`].
+    ///
+    /// The entry for a copy to is removed once at least one replica has finished.
```
"... or the exporting collection is dropped."

teskje reviewed (approved):
Compute changes LGTM.

moulimukherjee commented on src/adapter/src/coord/message_handler.rs:
```diff
@@ -343,12 +345,36 @@ impl Coordinator {
+            }
+            ControllerResponse::CopyToResponse(sink_id, response) => {
+                match self.active_compute_sinks.get_mut(&sink_id) {
+                    Some(ActiveComputeSink::CopyTo(active_copy_to)) => {
+                        let response = match response {
+                            Ok(n) => Ok(ExecuteResponse::Copied(usize::cast_from(n))),
+                            Err(error) => Err(AdapterError::Unstructured(error)),
+                        };
+                        active_copy_to.process_response(response);
+                        self.drop_compute_sinks([(sink_id, ComputeSinkRemovalReason::Finished)])
+                            .await;
```
@benesch This is after cherry-picking from https://github.com/MaterializeInc/materialize/pull/25260. 

If I remove the entry from `active_compute_sinks` by taking an owned value here (to make the changes for `ctx` to be not an option anymore) and process the response, then `drop_compute_sinks` will not actually clean up the dataflow.

moulimukherjee replied:
One option is, to maybe pass a parameterized `ComputeSinkRemovalReason::Finished`.

mjibson replied:
I'm ok with either 2 or 3 as long as it's really obvious which thing owns retiring the ctx. 3 sounds like it's more explicit so that'd be my first choice.

moulimukherjee replied:
Ack, let me do 3, and folks can weigh in. I think that would be cleaner as well.

benesch replied:
I'm not sure that'll wind up all that much clearer? Lemme start a discussion on Slack! https://materializeinc.slack.com/archives/C063MF1DAKH/p1708019372268269

moulimukherjee replied:
closing loop from slack https://materializeinc.slack.com/archives/C063MF1DAKH/p1708019372268269, going with 2 here. Do a follow up for a drop_compute_sinks refactor.

antiguru commented on src/compute-client/src/metrics.rs:
```diff
@@ -205,6 +213,8 @@ pub struct InstanceMetrics {
     /// TODO(#25239): Add documentation.
     pub subscribe_count: UIntGauge,
     /// TODO(#25239): Add documentation.
+    pub copy_to_count: UIntGauge,
+    /// TODO(#25239): Add documentation.
```
Please add documentation instead of copying the placeholder.

antiguru commented on src/compute/src/sink/s3_oneshot.rs:
```diff
@@ -27,18 +32,116 @@ where
+
+            ok_input.for_each(|_, data| {
+                data.swap(&mut ok_buf);
+                let count: u64 = ok_buf.drain(..).len().try_into().expect("usize into u64");
```
You'll need to filter data here for your up-to frontier, because there is no guarantee that a batch at time $t$ will only contain data at time $t$, but instead the guarantee is that forall data's time $t_d: t \leq t_d$. (This independent of what the system does and good habit for defensive programming.)

moulimukherjee replied:
Ack. I will clean this up in subsequent PRs where I will implement this sink properly. Currently I just needed the sink dataflow to return _some_ value for the response protocol to be defined.

moulimukherjee commented:
I have rebased the entire PR into 4 commits, which can be reviewed one by one if needed.

- https://github.com/MaterializeInc/materialize/pull/25042/commits/9993d217b5f862488547e582deccfea406b86838 adds the basic copy_to plumbing
- https://github.com/MaterializeInc/materialize/pull/25042/commits/d501f03cd870ff26da70e7191da2591dcad0b3fc is a refactor which combines `active_subscribes` and `active_copy_tos` to a common `active_compute_sinks`
- https://github.com/MaterializeInc/materialize/pull/25042/commits/919a593926508e378ca2a9bab491b63bb77c7fce This is cherry-picked https://github.com/MaterializeInc/materialize/pull/25260 and updated to compile
- https://github.com/MaterializeInc/materialize/pull/25042/commits/bca56ec299358bb386adcce3fb822bf68819986a minor refactors to also eagerly drop the compute sink after receiving a CopyToResponse + some additional comments.

moulimukherjee commented on src/adapter/src/copy_to.rs:
```diff
@@ -0,0 +1,45 @@
+#[derive(Debug)]
+pub struct ActiveCopyTo {
+    /// The user of the session that created the subscribe.
+    pub user: User,
```
(Resolving since we addressed this in slack)

benesch reviewed (approved):
LGTM! Thank you for slogging through another hefty change!

moulimukherjee commented:
Triggered nightly https://buildkite.com/materialize/nightlies/builds/6474#018dae3a-5883-49f4-8f3d-78a3344e1cff @nrainer-materialize I selected mostly which seemed relevant (this is a new `COPY TO`, which is behind a flag so not yet used, but the code path does touch existing `SUBSCRIBE`s)

moulimukherjee commented:
For the cargo test failure, will have to rebase from main again. Will let the nightlies run before doing that though, they are still mostly green.
